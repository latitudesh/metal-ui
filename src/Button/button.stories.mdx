import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import Button from "./index";
import Icon from "../../util/Icon";
import { ArrowUpOutline } from "heroicons-react";
import Box from "../Box";

<Meta title="Button" component={Button} />

# API

<Props of={Button} />

# Variants

Brand secondary is meant to be used against dark background.

<Preview>
  <Story name="Variants">
    <Box flex justifyContent="evenly">
      <Button label="Primary" />
      <Button label="Secondary" variant="secondary" />
      <Button label="Danger" variant="danger" />
      <Button label="Minimal" variant="minimal" />
      <Button label="Brand primary" variant="brand-p" />
      <div className="block px-2 py-2 -mt-2 bg-brand-deepPurple">
        <Button label="Brand secondary" variant="brand-s" />
      </div>
    </Box>
  </Story>
</Preview>

# Sizes

<Preview>
  <Story name="Size">
    <div>
      <Box flex justifyContent="evenly">
        <Button label="Primary" size="small" />
        <Button label="Default" />
        <Button label="Primary" size="large" />
      </Box>
      <Button label="Block small" size="small" block />
      <div className="mt-2">
        <Button label="Block normal" block />
      </div>
      <div className="mt-2">
        <Button label="Block large" size="large" block />
      </div>
    </div>
  </Story>
</Preview>

# Disabled

Disabled buttons do not inherit `variant` styling.

<Preview>
  <Story name="Disabled">
    <Box flex justifyContent="around">
      <Button label="Disabled" disabled />
      <Button
        label="Disabled"
        disabled
        iconBefore={<ArrowUpOutline size={20} />}
      />
      <Button
        label="Disabled"
        disabled
        size="large"
        iconBefore={<ArrowUpOutline size={20} />}
        iconAfter={<ArrowUpOutline size={20} />}
      />
    </Box>
  </Story>
</Preview>

# Loading

<Preview>
  <Story name="Loading">
    <Box flex justifyContent="around">
      <Button label="Loading" isLoading={true} />
      <Button label="Loading" variant="secondary" isLoading={true} />
      <Button label="Loading" variant="danger" isLoading={true} />
    </Box>
    <Box flex justifyContent="around">
      <Button label="Loading" block isLoading={true} />
    </Box>
  </Story>
</Preview>

# With Icon

The component does not handle icon sizing. You should manage that from your project. `20px` is usually the recommended size.

<Preview>
  <Story name="With Icon">
    <Box flex justifyContent="around">
      <Button iconBefore={<ArrowUpOutline size={20} />} label="Icon before" />
      <Button
        iconAfter={<ArrowUpOutline size={20} />}
        variant="secondary"
        label="Icon after"
      />
      <Button
        iconBefore={<ArrowUpOutline size={20} />}
        iconAfter={<ArrowUpOutline size={20} />}
        variant="secondary"
        label="Icon before and after"
      />
    </Box>
  </Story>
</Preview>

# With `next/link`

For use in conjunction with NextJS' `next/link`, wrap the button with `<Link>` and set `passHref` to true.

```javascript
<Link href="https://www.maxihost.com" passHref>
  <Button
    label="Go to website"
  />
</Link>
```
