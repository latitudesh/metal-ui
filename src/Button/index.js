/** @jsxRuntime classic */
/** @jsx jsx */
import { jsx } from "@emotion/react";
import React, { cloneElement } from "react";
import PropTypes from "prop-types";
import SpinningDots from "../SpinningDots";
import tw from "twin.macro";

const Button = React.forwardRef(({
  disabled,
  iconBefore,
  iconAfter,
  onClick,
  label,
  type,
  size,
  variant,
  block,
  isLoading,
  className,
}, ref) => {
  const ButtonContent = (
    <div
      tw="flex items-center justify-center"
      css={[isLoading ? tw`opacity-0` : tw`opacity-100`]}
    >
      {iconBefore && <span tw="flex mr-2">{cloneElement(iconBefore)}</span>}
      <span tw="truncate">{label}</span>
      {iconAfter && <span tw="flex ml-2">{cloneElement(iconAfter)}</span>}
    </div>
  );

  return (
    <button
      type={type}
      onClick={onClick}
      disabled={disabled || isLoading}
      className={className}
      ref={ref}
      css={[
        tw`relative border items-center flex font-medium rounded focus:outline-none focus:ring transition ease-in-out duration-150 justify-center max-w-full cursor-pointer`,
        disabled &&
          tw`border-border text-accent-four hover:text-accent-four active:text-accent-four bg-accent-two hover:bg-accent-two active:bg-accent-two cursor-not-allowed`,
        variant == "default" &&
          !disabled &&
          tw`border-transparent hover:border-accent-five text-white hover:text-foreground bg-foreground hover:bg-accent-two active:bg-foreground active:text-white`,
        variant == "secondary" &&
          !disabled &&
          tw`border-secondary hover:border-foreground text-secondary hover:text-foreground bg-white hover:bg-accent-two active:bg-foreground active:text-white`,
        variant == "danger" &&
          !disabled &&
          tw`border-error-dark hover:border-error-dark text-white hover:text-error-dark bg-error-dark hover:bg-accent-two active:bg-foreground active:text-white`,
        variant == "minimal" &&
          !disabled &&
          tw`border-transparent bg-transparent hover:bg-accent-three active:bg-accent-two text-foreground hover:text-accent-seven active:text-accent-five`,
        variant == "brand-p" &&
          !disabled &&
          tw`border-transparent text-white hover:text-white bg-brand-uv hover:bg-opacity-90 active:bg-opacity-90 active:text-white`,
        variant == "brand-s" &&
          !disabled &&
          tw`border-brand-gray text-white hover:text-white bg-transparent hover:opacity-80 active:text-white`,
        size == "small" && tw`px-4 h-8 leading-8 text-sm`,
        size == "normal" && tw`px-6 h-9 leading-9 text-sm`,
        size == "large" && tw`px-12 h-10 leading-10 text-base`,
        block && tw`w-full`,
      ]}
    >
      {isLoading && <SpinningDots variant="inherit" className={tw`absolute`} />}
      {ButtonContent}
    </button>
  );
});

Button.propTypes = {
  disabled: PropTypes.bool,
  className: PropTypes.string,
  iconBefore: PropTypes.element,
  iconAfter: PropTypes.element,
  label: PropTypes.string,
  onClick: PropTypes.func,
  type: PropTypes.string,
  variant: PropTypes.oneOf([
    "default",
    "secondary",
    "danger",
    "disabled",
    "minimal",
    "brand-p",
    "brand-s",
  ]),
  size: PropTypes.oneOf(["small", "normal", "large"]),
  isLoading: PropTypes.bool,
};

Button.defaultProps = {
  size: "normal",
  variant: "default",
  type: "button",
  isLoading: false,
  disabled: false,
};

export default Button;
