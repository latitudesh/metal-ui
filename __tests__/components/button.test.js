import React from "react";
import renderer from "react-test-renderer";

import Button from "../../src/Button";

describe("Snapshots/Components/Button", () => {
  it("Render button", () => {
    const component = renderer.create(<Button />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
