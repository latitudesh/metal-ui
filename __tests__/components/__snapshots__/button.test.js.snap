// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots/Components/Button Render button 1`] = `
.emotion-0 {
  transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform,filter,backdrop-filter;
  transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
  transition-duration: 150ms;
  position: relative;
  border-width: 1px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-weight: 500;
  border-radius: 5px;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  max-width: 100%;
  cursor: pointer;
  border-color: rgba(0, 0, 0, 0);
  --tw-text-opacity: 1;
  color: rgba(255, 255, 255, var(--tw-text-opacity));
  --tw-bg-opacity: 1;
  background-color: rgba(0, 0, 0, var(--tw-bg-opacity));
  padding-left: 1.5rem;
  padding-right: 1.5rem;
  height: 2.25rem;
  line-height: 1.25rem;
  font-size: 0.875rem;
}

.emotion-0:focus {
  outline: 2px solid transparent;
  outline-offset: 2px;
  --tw-ring-offset-shadow: var(--tw-ring-inset) 0 0 0 var(--tw-ring-offset-width) var(--tw-ring-offset-color);
  --tw-ring-shadow: var(--tw-ring-inset) 0 0 0 calc(3px + var(--tw-ring-offset-width)) var(--tw-ring-color);
  box-shadow: var(--tw-ring-offset-shadow),var(--tw-ring-shadow),var(--tw-shadow, 0 0 #0000);
}

.emotion-0:hover {
  --tw-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.1),0 1px 2px 0 rgba(0, 0, 0, 0.06);
  box-shadow: var(--tw-ring-offset-shadow, 0 0 #0000),var(--tw-ring-shadow, 0 0 #0000),var(--tw-shadow);
}

.emotion-0:hover,
.emotion-0:focus {
  --tw-bg-opacity: 1;
  background-color: rgba(17, 17, 17, var(--tw-bg-opacity));
  opacity: 0.95;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  opacity: 1;
}

.emotion-2 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<button
  className="emotion-0"
  disabled={false}
  type="button"
>
  <div
    className="emotion-1"
  >
    <span
      className="emotion-2"
    />
  </div>
</button>
`;
